qoriq: Add support for NXP T4240RDB board

CHKDSK88 opened the pull request:
Hardware specs:
  - NXP T4240, 12C/24T @ 1.67 GHz
  - 3 × 2 GB DDR3 SO-DIMM
  - 128 MB NOR flash
  - 2 GB SLC NAND
  - SD card interface
  - PCIe: x4 and x8
  - SATA 3 Gbps
  - 8 × 1 GbE
  - 4 × 10 GbE SFP
  - RTC

This commit adds the sysupgrade and factory images for T4240RDB board in
both variants:
  - nor: for booting and read whole system from NOR memory
  - sdboot: for booting and read whole system from SD card

SD Card images install:

  - Burn image to sdcard. E.g:
```
gunzip -c gunzip -c openwrt-qoriq-generic-fsl_T4240RDB-squashfs-sdcard.img.gz | \
sudo dd of=/dev/mmcblk0 conv=fsync,notrunc status=progress bs=4M && sync
```
  - Download lastest Cortina PHY firmware from NXP github [1], if you accept their
    EULA [2].
  - Install Cortina PHY on image, E.g:
`dd if=cs4315-cs4340-PHY-ucode.txt of=/dev/mmcblk0 bs=1 seek=2M`
  - Insert SD-Card to SD slot
  - Switch SW3.4 to OFF
  - Configre mac addresses from sticker in u-boot. E.g:
```
setenv ethaddr 00:10:f3:3a:a8:66
setenv eth1addr 00:10:f3:3a:a8:67
setenv eth2addr 00:10:f3:3a:a8:68
setenv eth3addr 00:10:f3:3a:a8:69
setenv eth4addr 00:10:f3:3a:a8:6a
setenv eth5addr 00:10:f3:3a:a8:6b
setenv eth6addr 00:10:f3:3a:a8:6c
setenv eth7addr 00:10:f3:3a:a8:6d
setenv eth8addr 00:10:f3:3a:a8:6e
setenv eth9addr 00:10:f3:3a:a8:6f
setenv eth10addr 00:10:f3:3a:a8:70
setenv eth11addr 00:10:f3:3a:a8:71
saveenv
```
  - reset and boot

NOR images install:

  - download and extract factory image on tftp server root
  - boot device and stop in u-boot (from nor or sd card u-boot)
  - configure server and ip address. E.g:
 ```
setenv ipaddr 192.168.1.2
setenv serverip 192.168.1.1
```
  - Download image and run flashing:
```
tftpboot $loadaddr openwrt-qoriq-generic-fsl_T4240RDB-squashfs-factory-nor.bin
protect off all
erase $fwaddr +$filesize
cp.b $loadaddr $fwaddr $filesize
```
  - Switch SW3.4 to ON
  - Switch SW3.1-3 to OFF
  - reboot
  - Do postprocessing (see bellow)

NOR images post processing:

  - Configre mac addresses from sticker in u-boot. E.g:
```
setenv ethaddr 00:10:f3:3a:a8:66
setenv eth1addr 00:10:f3:3a:a8:67
setenv eth2addr 00:10:f3:3a:a8:68
setenv eth3addr 00:10:f3:3a:a8:69
setenv eth4addr 00:10:f3:3a:a8:6a
setenv eth5addr 00:10:f3:3a:a8:6b
setenv eth6addr 00:10:f3:3a:a8:6c
setenv eth7addr 00:10:f3:3a:a8:6d
setenv eth8addr 00:10:f3:3a:a8:6e
setenv eth9addr 00:10:f3:3a:a8:6f
setenv eth10addr 00:10:f3:3a:a8:70
setenv eth11addr 00:10:f3:3a:a8:71
saveenv
```
  - boot
  - Download and refresh RCW stored in eeprom:
```
tr '\0' '\377' < /dev/zero | dd bs=256 of=/sys/bus/i2c/devices/0-0050/eeprom
cat /tmp/openwrt-qoriq-generic-fsl_T4240RDB-squashfs-rcw.bin > /sys/bus/i2c/devices/0-0050/eeprom
```
  - Download lastest Cortina PHY firmware from NXP github [1], if you accept their
    EULA [2].
  - Install Cortina PHY on image, E.g:
`mtd write cs4315-cs4340-PHY-ucode.txt /dev/mtd4`
  - reset and boot

[1] https://raw.githubusercontent.com/nxp-qoriq/qoriq-firmware-cortina/refs/tags/lf-6.12.34-2.1.0/cs4315-cs4340-PHY-ucode.txt
[2] https://github.com/nxp-qoriq/qoriq-firmware-cortina/blob/lf-6.12.34-2.1.0/EULA.txt

CHKDSK88 committed:
package: u-boot: initial support for qoriq arch

This package adds initial u-boot support for qoriq target.

U-boot for qoriq devices must be compiled with 32-bit compiler and
linked with 32-bit linker. It's part of mpc 85xx target. But qoriq
target is 64-bit. As workaround, mpc85xx binary toolchain is downloaded
only for this u-boot.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/10941
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

f00b4r0 commented:
@stintel ping :)

stintel commented:
I have cherry-picked and pushed a modified version of the first patch in this PR to OpenWrt master; thanks!

Due to the EULA on the Cortina PHY and FMAN firmwares, we most likely cannot ship them as part of our images. See e.g. [discussion on the ML about firmware-imx](http://lists.openwrt.org/pipermail/openwrt-devel/2022-August/039247.html). For the same reason, we should probably remove the whole fman-ucode package from our tree. Added this on the next developer meeting agenda. We will discuss there, and ask advice from SFC if needed. For now, the PR cannot be accepted as-is. Do we really need to those firmwares to be included in the image? On the M300, at least the FMAN firmware is on the NOR. If we can drop both these firmwares from the PR, it could be acceptable. The cortina-ucode and fman-ucode NOR partitions suggest we can go for this approach.

As for the vendor U-Boot fork: do we really need this? On the M300, a working U-Boot is on the NOR. I would prefer taking the same approach there, and avoid the vendor fork entirely. It would also avoid having to download a binary toolchain, which imo is a rather ugly hack. The u-boot NOR partition suggests we can go for this approach.

Is the custom RCW needed? If so, please move this to a separate commit, and explain why in the commit message.

Device Tree:
* please create a dtsi with common properties and include it in the nor and sdboot dts to avoid duplicate code
* drop the whole-nor and whole-nand partitions

Kernel config:
* if SATA is disabled in dts, do we really need to enable support for it in the kernel?
* disabling CONFIG_DPAA_ERRATUM_A050385 should ideally go in a separate commit explaining it is only relevant on ARM64 platforms - it is not really related to the T4240RDB, and might raise questions without explanation

MAC addresses: do we really need to manually set them? Aren't they in one of the NOR partitions?

Do we really need another gen_sdcard script? The script is identical to target/linux/layerscape/image/gen_sdcard_head_img.sh. Move it to the toplevel scripts directory instead, in a separate commit, and update its path in target/linux/layerscape/image/Makefile in the same commit to avoid breakage on bisecting.

stintel commented:
> MAC addresses: do we really need to manually set them? Aren't they in one of the NOR partitions?

If we really need to manually set them, consolidate the instructions in "common post processing" section or so?

And this is probably a stretch, but do you know where to obtain this device for an affordable price? 😃

CHKDSK88 commented:
@stintel 

Please me know which approach will be the best.

stintel commented:
Sorry, haven't found the time to go over this properly. I was actually looking at it earlier today, and got again stuck on CONFIG_PAHOLE_HAS_SPLIT_BTF reappearing in target/linux/qoriq/config-5.15. Even after introducing a config symbol for it like this:

```diff
--- a/config/Config-kernel.in
+++ b/config/Config-kernel.in
@@ -387,6 +387,11 @@ config KERNEL_DEBUG_INFO_REDUCED
          DEBUG_INFO build and compile times are reduced too.
          Only works with newer gcc versions.

+config KERNEL_PAHOLE_HAS_SPLIT_BTF
+       bool
+       default y
+       depends on KERNEL_DEBUG_INFO
+
 # KERNEL_DEBUG_LL symbols must have the default value set as otherwise
 # KConfig wont evaluate them unless KERNEL_EARLY_PRINTK is selected
 # which means that buildroot wont override the DEBUG_LL symbols in target
```

CHKDSK88 commented:
@stintel Could You explain why only qoriq target need it? I can't find this config on any other target.

stintel commented:
> @stintel Could You explain why only qoriq target need it? I can't find this config on any other target.

It's not that the qoriq target needs it, just enabling CONFIG_KERNEL_DEBUG_INFO or CONFIG_KERNEL_DEBUG_INFO_BTF seems to leak that symbol to the target kernel config, which shouldn't happen, as we have CONFIG_KERNEL_PAHOLE_HAS_SPLIT_BTF to handle that.

65a commented:
I have this board, and have been testing the patches here on NOR. I was getting the infamous "FD error", it turns out dpaa_eth has some issues with jumbo frames. I was runing a 6.3rc under gentoo, and I can't remember if I tried this. As a workaround, you can do:
```
diff --git a/drivers/net/ethernet/freescale/fman/fman.c b/drivers/net/ethernet/freescale/fman/fman.c
index ce0a121580f6..597374ac9a87 100644
--- a/drivers/net/ethernet/freescale/fman/fman.c
+++ b/drivers/net/ethernet/freescale/fman/fman.c
@@ -2425,7 +2425,7 @@ EXPORT_SYMBOL(fman_get_mem_region);
 #define FSL_FM_RX_EXTRA_HEADROOM_MAX	384

 /* Maximum frame length */
-#define FSL_FM_MAX_FRAME_SIZE			1522
+#define FSL_FM_MAX_FRAME_SIZE			9600
 #define FSL_FM_MAX_POSSIBLE_FRAME_SIZE		9600
 #define FSL_FM_MIN_POSSIBLE_FRAME_SIZE		64
```
There was another issue in u-boot's build scripts, basically it injects "u-boot-$VERSION" as the version, which python's versioning dislikes. I added:
`sed -i 's/u-boot-//g' $(PKG_BUILD_DIR)/scripts/dtc/pylibfdt/Makefile`  to the package Makefile to fix it, but I assume this is because u-boot is stale. I didn't actually use/need RCW or u-boot, so I haven't tested that.

Regardless of linux distro, I've had some issues getting this board to go 10GB/s, but it appears jumbo frames get me somewhat close in iperf3. I am still evaluating software bridge/switch performance, I wonder if there is a dpaa hardware way to do it, or dpdk stuff. I know this is possible under dpaa2.

Not sure if OpenWRT can support it, but a "rootfs on NOR, overlay on the 2GB NAND with ubifs" config would be  pretty neat.

Thanks for the patch, I'll keep playing around with it, let me know if there's anything you want me to test.

f00b4r0 commented:
@CHKDSK88 

> > Do we really need another gen_sdcard script? The script is identical to target/linux/layerscape/image/gen_sdcard_head_img.sh. Move it to the toplevel scripts directory instead, in a separate commit, and update its path in target/linux/layerscape/image/Makefile in the same commit to avoid breakage on bisecting.
> 
> I will clean it, but we first must discuss what approach we take.
> 
> I see couple ways for this device:
> 
> 1. We forgot about  Cortina uCode and keep factory images. (U-boot stays as is. It's ugly but complete solution.)
> 2. In mainline tree we keep only sysupgrade images and I prepare instruction for 'home-cooking' factory images if someone will need it.  (Something like P2020RDB support approach)
> 
> Please let me know which way I should take.

In order to improve the status of this PR, I would suggest going with 1. AIUI this removes all the currently questionable material from this PR, which is the current blocker here. The situation can always be revisited at a later time if the copyright/distributability issues are cleared.

My 2c.

neheb commented:
MAC addresses.

Does uboot set them?

CHKDSK88 committed:
qoriq: add support for NXP T4240RDB board

Hardware specs:
  - NXP T4240, 12C/24T @ 1.67 GHz
  - 3 × 2 GB DDR3 SO-DIMM
  - 128 MB NOR flash
  - 2 GB SLC NAND
  - SD card interface
  - PCIe: x4 and x8
  - SATA 3 Gbps
  - 8 × 1 GbE
  - 4 × 10 GbE SFP
  - RTC

This commit adds the sysupgrade and factory images for T4240RDB board in
both variants:
  - nor: for booting and read whole system from NOR memory
  - sdboot: for booting and read whole system from SD card

SD Card images install:

  - Burn image to sdcard. E.g:
      gunzip -c gunzip -c openwrt-qoriq-generic-fsl_T4240RDB-squashfs-sdcard.img.gz | \
      sudo dd of=/dev/mmcblk0 conv=fsync,notrunc status=progress bs=4M && sync
  - Download lastest Cortina PHY firmware from NXP github [1], if you accept their
    EULA [2].
  - Install Cortina PHY on image, E.g:
      dd if=cs4315-cs4340-PHY-ucode.txt of=/dev/mmcblk0 bs=1 seek=2M
  - Insert SD-Card to SD slot
  - Switch SW3.4 to OFF
  - Configre mac addresses from sticker in u-boot. E.g:
      setenv ethaddr 00:10:f3:3a:a8:66
      setenv eth1addr 00:10:f3:3a:a8:67
      setenv eth2addr 00:10:f3:3a:a8:68
      setenv eth3addr 00:10:f3:3a:a8:69
      setenv eth4addr 00:10:f3:3a:a8:6a
      setenv eth5addr 00:10:f3:3a:a8:6b
      setenv eth6addr 00:10:f3:3a:a8:6c
      setenv eth7addr 00:10:f3:3a:a8:6d
      setenv eth8addr 00:10:f3:3a:a8:6e
      setenv eth9addr 00:10:f3:3a:a8:6f
      setenv eth10addr 00:10:f3:3a:a8:70
      setenv eth11addr 00:10:f3:3a:a8:71
      saveenv
  - reset and boot

NOR images install:

  - download and extract factory image on tftp server root
  - boot device and stop in u-boot (from nor or sd card u-boot)
  - configure server and ip address. E.g:
      setenv ipaddr 192.168.1.2
      setenv serverip 192.168.1.1
  - Download image and run flashing:
      tftpboot $loadaddr openwrt-qoriq-generic-fsl_T4240RDB-squashfs-factory-nor.bin
      protect off all
      erase $fwaddr +$filesize
      cp.b $loadaddr $fwaddr $filesize
  - Switch SW3.4 to ON
  - Switch SW3.1-3 to OFF
  - reboot
  - Do postprocessing (see bellow)

NOR images post processing:

  - Configre mac addresses from sticker in u-boot. E.g:
      setenv ethaddr 00:10:f3:3a:a8:66
      setenv eth1addr 00:10:f3:3a:a8:67
      setenv eth2addr 00:10:f3:3a:a8:68
      setenv eth3addr 00:10:f3:3a:a8:69
      setenv eth4addr 00:10:f3:3a:a8:6a
      setenv eth5addr 00:10:f3:3a:a8:6b
      setenv eth6addr 00:10:f3:3a:a8:6c
      setenv eth7addr 00:10:f3:3a:a8:6d
      setenv eth8addr 00:10:f3:3a:a8:6e
      setenv eth9addr 00:10:f3:3a:a8:6f
      setenv eth10addr 00:10:f3:3a:a8:70
      setenv eth11addr 00:10:f3:3a:a8:71
      saveenv
  - boot
  - Download and refresh RCW stored in eeprom:
      tr '\0' '\377' < /dev/zero | dd bs=256 of=/sys/bus/i2c/devices/0-0050/eeprom
      cat /tmp/openwrt-qoriq-generic-fsl_T4240RDB-squashfs-rcw.bin > /sys/bus/i2c/devices/0-0050/eeprom
  - Download lastest Cortina PHY firmware from NXP github [1], if you accept their
    EULA [2].
  - Install Cortina PHY on image, E.g:
      mtd write cs4315-cs4340-PHY-ucode.txt /dev/mtd4
  - reset and boot

[1] https://raw.githubusercontent.com/nxp-qoriq/qoriq-firmware-cortina/refs/tags/lf-6.12.34-2.1.0/cs4315-cs4340-PHY-ucode.txt
[2] https://github.com/nxp-qoriq/qoriq-firmware-cortina/blob/lf-6.12.34-2.1.0/EULA.txt

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/10941
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

CHKDSK88 committed:
qoriq: kernel: refresh config

Done by 'make kernel_oldconfig'.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/10941
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

CHKDSK88 commented:
@65a 
> I have this board, and have been testing the patches here on NOR. I was getting the infamous "FD error", it turns out dpaa_eth has some issues with jumbo frames. I was runing a 6.3rc under gentoo, and I can't remember if I tried this. As a workaround, you can add the following to the `patches` directory under `targets/linux/qoriq`:
> 
> ```
> diff --git a/drivers/net/ethernet/freescale/fman/fman.c b/drivers/net/ethernet/freescale/fman/fman.c
> index ce0a121580f6..597374ac9a87 100644
> --- a/drivers/net/ethernet/freescale/fman/fman.c
> +++ b/drivers/net/ethernet/freescale/fman/fman.c
> @@ -2425,7 +2425,7 @@ EXPORT_SYMBOL(fman_get_mem_region);
>  #define FSL_FM_RX_EXTRA_HEADROOM_MAX	384
> 
>  /* Maximum frame length */
> -#define FSL_FM_MAX_FRAME_SIZE			1522
> +#define FSL_FM_MAX_FRAME_SIZE			9600
>  #define FSL_FM_MAX_POSSIBLE_FRAME_SIZE		9600
>  #define FSL_FM_MIN_POSSIBLE_FRAME_SIZE		64
> ```


It could be fixed via bootargs. Add to bootargs: `fsl_dpaa_fman.fsl_fm_max_frm=9600`. No patch will be required.

stintel commented:
The patch target/linux/qoriq/patches-6.12/001-Revert-powerpc-dts-mpc85xx-remove-simple-bus-compatible-from-ifc-node.patch seems copied from target/linux/mpc85xx/patches-6.12/002-Revert-powerpc-dts-mpc85xx-remove-simple-bus-compatible-from-ifc-node.patch. In this case, to avoid duplicates, it should be moved to target/linux/generic/pending-6.12, as it has not been accepted yet: https://patchwork.ozlabs.org/project/linuxppc-dev/patch/20251105205524.17362-1-rosenp@gmail.com/

CHKDSK88 committed:
kernel: move patch from mpc85xx to pending

The patch "Revert "powerpc: dts: mpc85xx: remove "simple-bus" compatible
from ifc node" has been sent upstream [0].

It is also required for qoriq target and in this commit it is moved to
the generic/pending folder.

[0] https://patchwork.ozlabs.org/project/linuxppc-dev/patch/20251105205524.17362-1-rosenp@gmail.com/

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/10941
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>
